mono/metadata/metadata.c

```diff
@@ -5276,6 +5291,21 @@ do_mono_metadata_type_equal (MonoType *t1, MonoType *t2, gboolean signature_only
 	if (t1->type != t2->type || t1->byref != t2->byref)
 		return FALSE;
 
+	if (t1->num_mods != t2->num_mods)
```

Comment:
Shouldn't this only be comparing required modifiers, not optional ones?

Reply:
Are two integers, one with an optional modifier, one without any modifiers, are they equal?

Reply:
https://blogs.msdn.microsoft.com/samng/2008/03/05/optional-modifiers-and-overload-resolution/ 

It looks like the main use of optional modifiers are to differentiate between functions taking a long and functions taking an int (or other bitness) on C++/CLI. Since they use the same underlying CLR type for both, they use modopts to prevent one function from being used where the other is expected. If we treat them as equal, that defeats that use.

Reply:
http://weblog.ikvm.net/2003/01/11/customModifiersAndReturnTypeCustomAttributes.aspx

Reply:
From 335, 7.1.1: 

```The CLI itself shall treat required and optional modifiers in the same manner.```